plugins/pulp_rpm/plugins/distributors/yum/metadata/repomd.py

```diff
@@ -44,6 +46,12 @@ def finalize(self):
             signer = util.Signer(options=self.sign_options)
             signer.sign(self.metadata_file_path)
 
+    def _keep_repomd(self, metadata_file_path):
```

Comment:
Just to be clear here, this is initialized in InitRepoMetadataStep, which means all files collected here can be outdated at end of publish. Consider following workflow:
1. Repo contains repodata older then threshold
2. User runs publish
3. file_locations will be filled with repodata from step 1.
4. new repodata files are created during publish process.
5. In CloseRepoMetadataStep new filenames will be written in new repomd
6. In RemoveOldRepodataStep no files will be removed because old files are kept in self.file_locations

So it would be better if you move this to CloseRepoMetadataStep::finalize

Reply:
Right I see what you're saying. Couldn't it also be done with moving the RemoveOldRepodataStep higher up in the .add_child hierarchy in the constructor for BaseYumRepoPublisher instead of adding it as the last child?

so the order would be:

PublishCompsStep
PublishMetadataStep
RemoveOldRepodataStep
CloseRepoMetadataStep
GenerateSqliteForRepoStep

Just an idea in case there's some case I haven't thought of that might cause this not to be a viable solution.

And just to clarify, you are just talking about moving the _keep_repomd and that's it to the CloseRepoMetadataStep class right? That's the way I am understanding it to fix the process.

Reply:
Yeah, moving to top of children hierarchy could solve the problem, but there you would have to check if repodata exists or if repository is published for first time.